File: go/wearable_biosignals.go
// wearable_biosignals.go
//
// Ports the CircleAI.Wearable.Biosignals module:
//   BiosignalKind (enum, explicit 0..8)          -> int consts, stable ordinals
//   BiosignalSample (record + Create factory)     -> value struct + constructor
//   IBiosignalSource (streaming contract)         -> BiosignalSource interface
//   NullBiosignalSource / RecordedBiosignalSource -> in-memory impls
//   BiosignalStats / BiosignalSnapshot (records)  -> value structs
//   BiosignalAggregator (sliding-window snapshot) -> struct
//   BiosignalAffectMapper (static rule sheet)     -> ApplyBiosignalToAffect
//
// Go mapping of the C# async surface:
//   IAsyncEnumerable<BiosignalSample> StreamAsync(ct)
//                            -> Stream(ctx) (<-chan BiosignalSample)
//   Task<bool> IsSupportedAsync(kind, ct)
//                            -> IsSupported(kind) bool  (the C# body is
//                               synchronous; the ct is unused there)
//   float (C#)               -> float32 (byte-identical rule thresholds)
//   Guid                     -> uuid.UUID
//
// CONCURRENCY: RecordedBiosignalSource.Stream creates the output channel and
// starts the producer goroutine; the goroutine closes the channel on completion
// or ctx cancellation, so a synchronous range over the channel terminates
// cleanly. BiosignalAggregator.Snapshot derives a window-bounded context so a
// never-completing source still yields a snapshot when the window elapses.
//
// The mapper mutates the shared AffectState (float32 fields), reusing the
// package clamp32 helper so results match the C# Math.Clamp(_, 0, 1) exactly.

package circleai

import (
	"context"
	"errors"
	"math"
	"time"

	"github.com/google/uuid"
)

// BiosignalKind is the canonical taxonomy of biosignals. Ports the BiosignalKind
// enum; integer values are explicit and stable (do not renumber).
type BiosignalKind int

const (
	// BiosignalKindHeartRate is heart rate, beats per minute.
	BiosignalKindHeartRate BiosignalKind = 0
	// BiosignalKindHeartRateVariability is HRV, RMSSD in milliseconds.
	BiosignalKindHeartRateVariability BiosignalKind = 1
	// BiosignalKindOxygenSaturation is SpO2, percent (0-100).
	BiosignalKindOxygenSaturation BiosignalKind = 2
	// BiosignalKindAccelerometer is accelerometer magnitude, m/s^2.
	BiosignalKindAccelerometer BiosignalKind = 3
	// BiosignalKindBodyTemperature is body temperature, degrees Celsius.
	BiosignalKindBodyTemperature BiosignalKind = 4
	// BiosignalKindSleepStage is sleep stage (0=awake,1=light,2=deep,3=REM).
	BiosignalKindSleepStage BiosignalKind = 5
	// BiosignalKindSteps is step count (cumulative or delta per IsCumulative).
	BiosignalKindSteps BiosignalKind = 6
	// BiosignalKindGalvanicSkinResponse is GSR, microsiemens.
	BiosignalKindGalvanicSkinResponse BiosignalKind = 7
	// BiosignalKindUnknown is a catch-all for vendor-specific/future signals.
	BiosignalKindUnknown BiosignalKind = 8
)

// BiosignalSample is a single biosignal measurement. Ports the BiosignalSample
// record. Value and Confidence are float32 (C# float).
type BiosignalSample struct {
	Id           uuid.UUID
	Kind         BiosignalKind
	Value        float32
	Unit         string
	Confidence   float32
	IsCumulative bool
	MeasuredAt   time.Time
}

// NewBiosignalSample creates a fresh sample with a new UUID, current UTC time,
// and confidence clamped to [0,1]. Ports BiosignalSample.Create (confidence
// defaults to 1.0, isCumulative to false at the C# call sites).
func NewBiosignalSample(kind BiosignalKind, value float32, unit string, confidence float32, isCumulative bool) BiosignalSample {
	return BiosignalSample{
		Id:           uuid.New(),
		Kind:         kind,
		Value:        value,
		Unit:         unit,
		Confidence:   clamp32(confidence, 0, 1),
		IsCumulative: isCumulative,
		MeasuredAt:   time.Now().UTC(),
	}
}

// BiosignalSource is a streaming source of biosignal samples. Ports
// IBiosignalSource.
type BiosignalSource interface {
	// SupportedKinds lists the kinds this source can emit (may be empty).
	SupportedKinds() []BiosignalKind
	// Stream emits samples until ctx is cancelled or the source is exhausted. The
	// returned channel is closed when streaming ends.
	Stream(ctx context.Context) <-chan BiosignalSample
	// IsSupported reports whether the source can produce the given kind.
	IsSupported(kind BiosignalKind) bool
}

// NullBiosignalSource supports nothing and emits nothing. Ports
// NullBiosignalSource ("no wearable connected").
type NullBiosignalSource struct{}

// SupportedKinds returns an empty slice. Ports SupportedKinds.
func (NullBiosignalSource) SupportedKinds() []BiosignalKind { return []BiosignalKind{} }

// IsSupported always returns false. Ports IsSupportedAsync.
func (NullBiosignalSource) IsSupported(kind BiosignalKind) bool { return false }

// Stream returns an immediately-closed channel. Ports StreamAsync (yield break).
func (NullBiosignalSource) Stream(ctx context.Context) <-chan BiosignalSample {
	ch := make(chan BiosignalSample)
	close(ch)
	return ch
}

// RecordedBiosignalSource replays a fixed list of samples. Ports
// RecordedBiosignalSource.
type RecordedBiosignalSource struct {
	samples     []BiosignalSample
	kinds       []BiosignalKind
	replayDelay time.Duration
}

// NewRecordedBiosignalSource constructs a replay source. Ports the constructor
// (nil samples -> error; replayDelay defaults to zero). SupportedKinds are the
// distinct kinds present in samples.
func NewRecordedBiosignalSource(samples []BiosignalSample, replayDelay time.Duration) (*RecordedBiosignalSource, error) {
	if samples == nil {
		return nil, errors.New("samples required")
	}
	seen := make(map[BiosignalKind]struct{})
	kinds := make([]BiosignalKind, 0)
	for _, s := range samples {
		if _, ok := seen[s.Kind]; !ok {
			seen[s.Kind] = struct{}{}
			kinds = append(kinds, s.Kind)
		}
	}
	cp := make([]BiosignalSample, len(samples))
	copy(cp, samples)
	return &RecordedBiosignalSource{samples: cp, kinds: kinds, replayDelay: replayDelay}, nil
}

// SupportedKinds lists the distinct kinds in the recording. Ports SupportedKinds.
func (s *RecordedBiosignalSource) SupportedKinds() []BiosignalKind {
	out := make([]BiosignalKind, len(s.kinds))
	copy(out, s.kinds)
	return out
}

// IsSupported reports whether the given kind appears in the recording. Ports
// IsSupportedAsync.
func (s *RecordedBiosignalSource) IsSupported(kind BiosignalKind) bool {
	for _, k := range s.kinds {
		if k == kind {
			return true
		}
	}
	return false
}

// Stream replays the recorded samples, honouring the optional replay delay and
// ctx cancellation. Ports StreamAsync.
func (s *RecordedBiosignalSource) Stream(ctx context.Context) <-chan BiosignalSample {
	ch := make(chan BiosignalSample)
	go func() {
		defer close(ch)
		for _, sample := range s.samples {
			if ctx.Err() != nil {
				return
			}
			if s.replayDelay > 0 {
				select {
				case <-ctx.Done():
					return
				case <-time.After(s.replayDelay):
				}
			}
			select {
			case <-ctx.Done():
				return
			case ch <- sample:
			}
		}
	}()
	return ch
}

// BiosignalStats is per-kind aggregate statistics over a window. Ports the
// BiosignalStats record.
type BiosignalStats struct {
	SampleCount int
	Min         float32
	Max         float32
	Mean        float32
}

// BiosignalSnapshot is a snapshot of aggregates across observed kinds. Ports the
// BiosignalSnapshot record. Kinds with no in-window samples are absent from
// Stats.
type BiosignalSnapshot struct {
	Stats       map[BiosignalKind]BiosignalStats
	GeneratedAt time.Time
}

// BiosignalAggregator computes sliding-window aggregates over a BiosignalSource.
// Ports BiosignalAggregator.
type BiosignalAggregator struct {
	source BiosignalSource
}

// NewBiosignalAggregator wraps the given source. Ports the constructor (nil
// source -> error).
func NewBiosignalAggregator(source BiosignalSource) (*BiosignalAggregator, error) {
	if source == nil {
		return nil, errors.New("source required")
	}
	return &BiosignalAggregator{source: source}, nil
}

// biosignalAccumulator accumulates count/min/max/sum for one kind.
type biosignalAccumulator struct {
	count int
	min   float32
	max   float32
	sum   float64
}

func newBiosignalAccumulator() *biosignalAccumulator {
	return &biosignalAccumulator{min: float32(math.Inf(1)), max: float32(math.Inf(-1))}
}

func (a *biosignalAccumulator) add(v float32) {
	a.count++
	if v < a.min {
		a.min = v
	}
	if v > a.max {
		a.max = v
	}
	a.sum += float64(v)
}

func (a *biosignalAccumulator) toStats() BiosignalStats {
	mean := float32(0)
	if a.count != 0 {
		mean = float32(a.sum / float64(a.count))
	}
	return BiosignalStats{SampleCount: a.count, Min: a.min, Max: a.max, Mean: mean}
}

// Snapshot consumes samples until the source completes or the window elapses,
// then returns aggregates over the in-window samples. Ports SnapshotAsync
// (window <= 0 -> error). Single-shot, not continuous.
func (agg *BiosignalAggregator) Snapshot(ctx context.Context, window time.Duration) (BiosignalSnapshot, error) {
	if window <= 0 {
		return BiosignalSnapshot{}, errors.New("Window must be positive.")
	}
	generatedAt := time.Now().UTC()
	cutoff := generatedAt.Add(-window)
	deadline := generatedAt.Add(window)

	// Time-bound the read so a never-completing source still yields a snapshot.
	wctx, cancel := context.WithTimeout(ctx, window)
	defer cancel()

	acc := make(map[BiosignalKind]*biosignalAccumulator)
	for sample := range agg.source.Stream(wctx) {
		if sample.MeasuredAt.Before(cutoff) {
			continue
		}
		a, ok := acc[sample.Kind]
		if !ok {
			a = newBiosignalAccumulator()
			acc[sample.Kind] = a
		}
		a.add(sample.Value)
		if !time.Now().UTC().Before(deadline) {
			break
		}
	}

	stats := make(map[BiosignalKind]BiosignalStats, len(acc))
	for kind, a := range acc {
		stats[kind] = a.toStats()
	}
	return BiosignalSnapshot{Stats: stats, GeneratedAt: generatedAt}, nil
}

// biosignalMinConfidence is the confidence gate below which samples never mutate
// affect. Ports BiosignalAffectMapper.MinConfidence.
const biosignalMinConfidence float32 = 0.5

// ApplyBiosignalToAffect applies the deterministic rule for sample to affect,
// mutating affect in place. Ports BiosignalAffectMapper.Apply. All mutated
// fields are clamped to [0,1]. Low-confidence samples are ignored.
//
// Rule sheet (clamped to [0,1]):
//   HeartRate  > 130 bpm: Energy += 0.10, Uncertainty += 0.05
//   HeartRate  > 100 bpm: Energy += 0.05
//   HeartRate  <  50 bpm: Energy -= 0.05
//   HRV        <  20 ms:  Uncertainty += 0.05, Rapport -= 0.02
//   HRV        >  60 ms:  Engagement += 0.02
//   SpO2       <  90 %:   Uncertainty += 0.10
//   SleepStage / others:  no mutation
func ApplyBiosignalToAffect(sample BiosignalSample, affect *AffectState) {
	if affect == nil {
		return
	}
	if sample.Confidence < biosignalMinConfidence {
		return
	}
	switch sample.Kind {
	case BiosignalKindHeartRate:
		applyHeartRate(sample.Value, affect)
	case BiosignalKindHeartRateVariability:
		applyHrv(sample.Value, affect)
	case BiosignalKindOxygenSaturation:
		applySpO2(sample.Value, affect)
	default:
		// SleepStage and the remaining kinds do not currently drive affect.
	}
	affect.LastUpdatedUTC = time.Now().UTC()
}

func applyHeartRate(bpm float32, a *AffectState) {
	switch {
	case bpm > 130:
		a.Energy = clamp32(a.Energy+0.10, 0, 1)
		a.Uncertainty = clamp32(a.Uncertainty+0.05, 0, 1)
	case bpm > 100:
		a.Energy = clamp32(a.Energy+0.05, 0, 1)
	case bpm < 50:
		a.Energy = clamp32(a.Energy-0.05, 0, 1)
	}
}

func applyHrv(rmssdMs float32, a *AffectState) {
	switch {
	case rmssdMs < 20:
		a.Uncertainty = clamp32(a.Uncertainty+0.05, 0, 1)
		a.Rapport = clamp32(a.Rapport-0.02, 0, 1)
	case rmssdMs > 60:
		a.Engagement = clamp32(a.Engagement+0.02, 0, 1)
	}
}

func applySpO2(percent float32, a *AffectState) {
	if percent < 90 {
		a.Uncertainty = clamp32(a.Uncertainty+0.10, 0, 1)
	}
}

// Interface guards.
var (
	_ BiosignalSource = NullBiosignalSource{}
	_ BiosignalSource = (*RecordedBiosignalSource)(nil)
)

File: rust/src/wearable_biosignals.rs
//! wearable_biosignals — CircleAI wearable biosignal layer.
//!
//! Full Rust port of `src/CircleAI.Wearable.Biosignals/`:
//!
//! - [`BiosignalKind`] (stable integer taxonomy 0–8), [`BiosignalSample`]
//!   (+ [`BiosignalSample::create`] factory), the [`IBiosignalSource`] contract,
//!   the [`NullBiosignalSource`] and [`RecordedBiosignalSource`] backends, the
//!   sliding-window [`BiosignalAggregator`] (+ [`BiosignalStats`] /
//!   [`BiosignalSnapshot`]), and the deterministic [`BiosignalAffectMapper`].
//!
//! The C# async surface is projected sync-only, mirroring the existing
//! `companion::IBioSignalStream` convention in this crate: `StreamAsync`
//! (`IAsyncEnumerable`) becomes [`IBiosignalSource::stream`] returning a
//! materialised `Vec`, and `IsSupportedAsync` (`Task<bool>`) becomes
//! [`IBiosignalSource::is_supported`]. `RecordedBiosignalSource` keeps its
//! `replay_delay` for parity but replays synchronously without sleeping.
//! `BiosignalAggregator::snapshot` drains the whole source once and aggregates
//! samples at/after `now - window` (there is no wall-clock deadline in the sync
//! projection — a finite source always terminates).

use std::collections::{HashMap, HashSet};

use chrono::{DateTime, Duration, Utc};
use uuid::Uuid;

use crate::memory::AffectState;

/// (Biosignals) Canonical kinds of biosignal sample. Integer values are stable
/// across language ports — do not renumber.
///
/// Mirrors `enum BiosignalKind`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum BiosignalKind {
    /// Heart rate, beats per minute.
    HeartRate = 0,
    /// Heart rate variability, RMSSD in milliseconds.
    HeartRateVariability = 1,
    /// Peripheral oxygen saturation, percent (0-100).
    OxygenSaturation = 2,
    /// Accelerometer magnitude, m/s^2.
    Accelerometer = 3,
    /// Body temperature, degrees Celsius.
    BodyTemperature = 4,
    /// Sleep stage encoded as a float: 0=awake, 1=light, 2=deep, 3=REM.
    SleepStage = 5,
    /// Step count (cumulative or delta — see [`BiosignalSample::is_cumulative`]).
    Steps = 6,
    /// Galvanic skin response, microsiemens.
    GalvanicSkinResponse = 7,
    /// Catch-all for vendor-specific or future signals.
    Unknown = 8,
}

/// (Biosignals) A single biosignal measurement.
///
/// Mirrors `sealed record BiosignalSample(Guid Id, BiosignalKind Kind,
/// float Value, string Unit, float Confidence, bool IsCumulative,
/// DateTimeOffset MeasuredAt)`.
#[derive(Debug, Clone, PartialEq)]
pub struct BiosignalSample {
    pub id: Uuid,
    pub kind: BiosignalKind,
    pub value: f32,
    pub unit: String,
    pub confidence: f32,
    pub is_cumulative: bool,
    pub measured_at: DateTime<Utc>,
}

impl BiosignalSample {
    /// Constructs a sample, mirroring the positional C# record constructor.
    pub fn new(
        id: Uuid,
        kind: BiosignalKind,
        value: f32,
        unit: impl Into<String>,
        confidence: f32,
        is_cumulative: bool,
        measured_at: DateTime<Utc>,
    ) -> Self {
        Self {
            id,
            kind,
            value,
            unit: unit.into(),
            confidence,
            is_cumulative,
            measured_at,
        }
    }

    /// Creates a fresh sample with a new UUID, current UTC timestamp, and
    /// confidence clamped to `[0, 1]`.
    ///
    /// Mirrors `BiosignalSample.Create(kind, value, unit, confidence = 1.0f,
    /// isCumulative = false)`.
    pub fn create(kind: BiosignalKind, value: f32, unit: impl Into<String>, confidence: f32, is_cumulative: bool) -> Self {
        Self {
            id: Uuid::new_v4(),
            kind,
            value,
            unit: unit.into(),
            confidence: confidence.clamp(0.0, 1.0),
            is_cumulative,
            measured_at: Utc::now(),
        }
    }
}

/// (Biosignals) A streaming source of biosignal samples.
///
/// Mirrors `interface IBiosignalSource` (sync-projected).
pub trait IBiosignalSource {
    /// The kinds this source can emit. May be empty for the null source.
    fn supported_kinds(&self) -> Vec<BiosignalKind>;
    /// The source's samples, in order (the sync analogue of `StreamAsync`).
    fn stream(&self) -> Vec<BiosignalSample>;
    /// Whether this source can produce samples of `kind`.
    fn is_supported(&self, kind: BiosignalKind) -> bool;
}

/// (Biosignals) A source that supports nothing and emits nothing.
///
/// Mirrors `sealed class NullBiosignalSource` — the "no wearable connected" case.
pub struct NullBiosignalSource;

impl NullBiosignalSource {
    /// Creates the null source.
    pub fn new() -> Self {
        Self
    }
}

impl Default for NullBiosignalSource {
    fn default() -> Self {
        Self::new()
    }
}

impl IBiosignalSource for NullBiosignalSource {
    fn supported_kinds(&self) -> Vec<BiosignalKind> {
        Vec::new()
    }
    fn stream(&self) -> Vec<BiosignalSample> {
        Vec::new()
    }
    fn is_supported(&self, _kind: BiosignalKind) -> bool {
        false
    }
}

/// (Biosignals) Replays a recorded biosignal stream.
///
/// Mirrors `sealed class RecordedBiosignalSource`. `replay_delay` is retained for
/// parity with the C# constructor but is not used to pace the sync replay.
pub struct RecordedBiosignalSource {
    samples: Vec<BiosignalSample>,
    kinds: Vec<BiosignalKind>,
    replay_delay: Duration,
}

impl RecordedBiosignalSource {
    /// Creates a source replaying `samples`. `replay_delay` defaults to zero when
    /// `None` (mirrors the C# `TimeSpan? replayDelay = null`).
    pub fn new(samples: Vec<BiosignalSample>, replay_delay: Option<Duration>) -> Self {
        // Distinct kinds, first-seen order (matches the C# HashSet enumerate).
        let mut seen: HashSet<BiosignalKind> = HashSet::new();
        let mut kinds: Vec<BiosignalKind> = Vec::new();
        for s in &samples {
            if seen.insert(s.kind) {
                kinds.push(s.kind);
            }
        }
        Self {
            samples,
            kinds,
            replay_delay: replay_delay.unwrap_or_else(Duration::zero),
        }
    }

    /// The configured replay delay (parity accessor; zero unless set).
    pub fn replay_delay(&self) -> Duration {
        self.replay_delay
    }
}

impl IBiosignalSource for RecordedBiosignalSource {
    fn supported_kinds(&self) -> Vec<BiosignalKind> {
        self.kinds.clone()
    }
    fn stream(&self) -> Vec<BiosignalSample> {
        self.samples.clone()
    }
    fn is_supported(&self, kind: BiosignalKind) -> bool {
        self.kinds.contains(&kind)
    }
}

/// (Biosignals) Per-kind aggregate statistics over a window.
///
/// Mirrors `sealed record BiosignalStats(int SampleCount, float Min, float Max,
/// float Mean)`.
#[derive(Debug, Clone, PartialEq)]
pub struct BiosignalStats {
    pub sample_count: i32,
    pub min: f32,
    pub max: f32,
    pub mean: f32,
}

impl BiosignalStats {
    /// Constructs stats, mirroring the positional C# record constructor.
    pub fn new(sample_count: i32, min: f32, max: f32, mean: f32) -> Self {
        Self {
            sample_count,
            min,
            max,
            mean,
        }
    }
}

/// (Biosignals) A point-in-time snapshot of per-kind aggregates.
///
/// Mirrors `sealed record BiosignalSnapshot(
/// IReadOnlyDictionary<BiosignalKind, BiosignalStats> Stats,
/// DateTimeOffset GeneratedAt)`.
#[derive(Debug, Clone, PartialEq)]
pub struct BiosignalSnapshot {
    pub stats: HashMap<BiosignalKind, BiosignalStats>,
    pub generated_at: DateTime<Utc>,
}

impl BiosignalSnapshot {
    /// Constructs a snapshot, mirroring the positional C# record constructor.
    pub fn new(stats: HashMap<BiosignalKind, BiosignalStats>, generated_at: DateTime<Utc>) -> Self {
        Self { stats, generated_at }
    }
}

/// Running min/max/mean accumulator — mirrors the private C# `Accumulator`.
struct Accumulator {
    count: i32,
    min: f32,
    max: f32,
    sum: f64,
}

impl Accumulator {
    fn new() -> Self {
        Self {
            count: 0,
            min: f32::INFINITY,
            max: f32::NEG_INFINITY,
            sum: 0.0,
        }
    }

    fn add(&mut self, v: f32) {
        self.count += 1;
        if v < self.min {
            self.min = v;
        }
        if v > self.max {
            self.max = v;
        }
        self.sum += v as f64;
    }

    fn to_stats(&self) -> BiosignalStats {
        let mean = if self.count == 0 {
            0.0
        } else {
            (self.sum / self.count as f64) as f32
        };
        BiosignalStats::new(self.count, self.min, self.max, mean)
    }
}

/// (Biosignals) Sliding-window aggregator over an [`IBiosignalSource`].
///
/// Mirrors `sealed class BiosignalAggregator`.
pub struct BiosignalAggregator<'a> {
    source: &'a dyn IBiosignalSource,
}

impl<'a> BiosignalAggregator<'a> {
    /// Wraps `source`.
    pub fn new(source: &'a dyn IBiosignalSource) -> Self {
        Self { source }
    }

    /// Drains the source once and returns a snapshot over the samples measured at
    /// or after `now - window`. Panics when `window <= 0` (mirrors the C#
    /// `ArgumentOutOfRangeException`).
    pub fn snapshot(&self, window: Duration) -> BiosignalSnapshot {
        if window <= Duration::zero() {
            panic!("Window must be positive.");
        }
        let generated_at = Utc::now();
        let cutoff = generated_at - window;
        let mut accumulator: HashMap<BiosignalKind, Accumulator> = HashMap::new();
        for sample in self.source.stream() {
            if sample.measured_at < cutoff {
                continue;
            }
            accumulator
                .entry(sample.kind)
                .or_insert_with(Accumulator::new)
                .add(sample.value);
        }
        let stats: HashMap<BiosignalKind, BiosignalStats> =
            accumulator.iter().map(|(k, acc)| (*k, acc.to_stats())).collect();
        BiosignalSnapshot::new(stats, generated_at)
    }
}

/// (Biosignals) Deterministic projection of biosignal samples onto
/// [`AffectState`] mutations.
///
/// Mirrors `static class BiosignalAffectMapper`. Rule sheet (all mutations
/// clamped to `[0, 1]`; confidence `< 0.5` never mutates):
///
/// - HeartRate `> 130` bpm: energy += 0.10, uncertainty += 0.05.
/// - HeartRate `> 100` bpm: energy += 0.05.
/// - HeartRate `< 50` bpm: energy -= 0.05.
/// - HRV `< 20` ms: uncertainty += 0.05, rapport -= 0.02.
/// - HRV `> 60` ms: engagement += 0.02.
/// - SpO2 `< 90` %: uncertainty += 0.10.
/// - SleepStage / other kinds: no mutation.
pub struct BiosignalAffectMapper;

impl BiosignalAffectMapper {
    const MIN_CONFIDENCE: f32 = 0.5;

    /// Applies the rule for `sample` to `affect`, mutating it in place. Stamps
    /// `last_updated_at` on any applicable (confidence-passing) sample, matching
    /// the C# which sets `LastUpdatedUtc` after the switch.
    pub fn apply(sample: &BiosignalSample, affect: &mut AffectState) {
        // Confidence gate — low-confidence samples never mutate state.
        if sample.confidence < Self::MIN_CONFIDENCE {
            return;
        }

        match sample.kind {
            BiosignalKind::HeartRate => Self::apply_heart_rate(sample.value, affect),
            BiosignalKind::HeartRateVariability => Self::apply_hrv(sample.value, affect),
            BiosignalKind::OxygenSaturation => Self::apply_spo2(sample.value, affect),
            // SleepStage and the remaining kinds do not drive affect.
            _ => {}
        }

        affect.last_updated_at = Utc::now();
    }

    fn apply_heart_rate(bpm: f32, a: &mut AffectState) {
        if bpm > 130.0 {
            a.energy = Self::clamp01(a.energy + 0.10);
            a.uncertainty = Self::clamp01(a.uncertainty + 0.05);
        } else if bpm > 100.0 {
            a.energy = Self::clamp01(a.energy + 0.05);
        } else if bpm < 50.0 {
            a.energy = Self::clamp01(a.energy - 0.05);
        }
    }

    fn apply_hrv(rmssd_ms: f32, a: &mut AffectState) {
        if rmssd_ms < 20.0 {
            a.uncertainty = Self::clamp01(a.uncertainty + 0.05);
            a.rapport = Self::clamp01(a.rapport - 0.02);
        } else if rmssd_ms > 60.0 {
            a.engagement = Self::clamp01(a.engagement + 0.02);
        }
    }

    fn apply_spo2(percent: f32, a: &mut AffectState) {
        if percent < 90.0 {
            a.uncertainty = Self::clamp01(a.uncertainty + 0.10);
        }
    }

    fn clamp01(v: f32) -> f32 {
        v.clamp(0.0, 1.0)
    }
}
